lib/jwt/jwa/signing_algorithm.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+module JWT
+  module JWA
+    module SigningAlgorithm
+      module ClassMethods
+        def register_algorithm(algo)
+          ::JWT::JWA.register_algorithm(algo)
+        end
+      end
+
+      def self.included(klass)
+        klass.extend(ClassMethods)
+      end
+
+      attr_reader :alg
+
+      def valid_alg?(alg_to_check)
+        alg&.casecmp(alg_to_check)&.zero? == true
+      end
+
+      def header(*)
+        { 'alg' => alg }
+      end
+
+      def raise_verify_error!(message)
+        raise(DecodeError.new(message).tap { |e| e.set_backtrace(caller(1)) })
+      end
+
+      def raise_sign_error!(message)
+        raise(EncodeError.new(message).tap { |e| e.set_backtrace(caller(1)) })
+      end
+    end
+
+    require_relative 'unsupported' # Require the unsupported algo as it's needed as a default for the rest
+
+    class << self
+      def register_algorithm(algo)
+        algorithms[algo.alg.to_s.downcase] = algo
+      end
+
+      def find(algo)
+        algorithms[algo.to_s.downcase]
+      end
+
+      private
+
+      def algorithms
+        @algorithms ||= {}.tap { |h| h.default = Unsupported }
```

Comment:
We could replace the use of `default` with a fallback to the key lookup in `find`. To avoid this strange require in the middle of everything.